final-project/application.py

```diff
@@ -25,103 +25,177 @@ def after_request(response):
+    def search_product_database(product_name_parameter):
+        try:
+            #Search the database for the requested product.
+            product_data = db.execute("SELECT *"
+                + " FROM products"
+                + " WHERE product_name"
+                + " LIKE :product"
+                + " LIMIT 1",
+                product = product_name_parameter)
 
+        except RuntimeError:
+            # If problem with db.execute, then apologize.
+            return apology("Error: We'll fix this. Please try again shortly.")
 
-@app.route("/new")
+        # Make sure we found a product.
+        if not product_data:
+            # If no products were found, then tell the user.
+            return apology("We do not have any products named "
+                + request_product() )
+
+        # Otherwise, return information about the requested product.
+        return product_data
+
+
+    # Assign local variables.
+    product_info = (
+        search_product_database(
+        make_parameters_flexible(
+        request_product() ) ) )
+    product_name = product_info[0]["product_name"]
+    image = product_info[0]["image"]
+    link = product_info[0]["link"]
+    description = product_info[0]["description"]
+    brand = product_info[0]["brand"]
+    price = usd(product_info[0]["price"])
+
+    #Render the product page.
+    return render_template("product.html",
+                            product_name=product_name,
+                            image=image,
+                            link=link,
+                            description=description,
+                            brand=brand,
+                            price=price)
+
+
+@app.route("/new", methods=["GET", "POST"])
 def new():
     """Add new product"""
 
     # remember user
     session.get("user_id")
 
+    # If user submits data for a new product, insert it into our table
     if request.method == "POST":
 
-        try:
-            # Add new product
-            # TODO: add request.form.get() to get parameters
-            new_product_row = db.execute("INSERT INTO products (id, category_id, product_name, link, description, image, brand, price) VALUES (NULL, 1, :product_name, :link, :description, :image, :brand, :price)", product_name = request.form.get("product_name"), link = request.form.get("link"), description = request.form.get("description"), image = request.form.get("image"), brand = request.form.get("brand"), price = request.form.get("price"))
-
-            # If a table constraint was violated
+        # Assign local variables
+        product_name=request.form.get("product_name")
+        link=request.form.get("link")
+        description=request.form.get("description")
+        image=request.form.get("image")
+        brand=request.form.get("brand")
+        price=request.form.get("price")
+
+        def save_products():
+            try:
+                # Save the new product in our products table.
+                new_product_row = db.execute(
```

Comment:
Are you sure db.execute throws? I'd expect it to just return Nones if the query doesn't return any results.

Reply:
Here is a link to the documentation. It is a quick read https://docs.cs50.net/problems/finance/finance.html#hints in easy language.

It looks like db.execute would only throw this type of error in a very rare circumstance. I could most likely get away without it, but I would rather be safe if I can. 

I might be able to create a helper method to run db.execute. For example, I could create a `run_sql(query):` method in `helper.py` then call it from `applications.py`. Woudl that be preferable? The try catches would still get run the same number of times, but they would be abstracted away and less repetitive in the code.

The ideal alternative is to switch to [SQLAlchemy Core] (http://docs.sqlalchemy.org/en/latest/index.html) or [Flask-SQLAlchemy](http://flask-sqlalchemy.pocoo.org/) (i.e., [SQLAlchemy ORM](http://docs.sqlalchemy.org/en/latest/index.html)) instead of `cs50.SQL`. But I would rather finish the MVP first.

Reply:
I think you should not try to catch the error. Anything that comes up is likely programmer error, and anything else should still be thrown and handled as a proper 500.

You can handle 500s without wrapping every endpoint call.

Reply:
Alright, I am going to make "handle 500 errors" a backlog item. I know they are uber important and one should never release software to users without handling them. However, since users are guaranteed not to run into these errors if the application is not at MVP, I am going to prioritize the work which prevents users from using the application first. 

It's definitely a good point though. I have to do more research on how to efficiently handle these. I searched around and it seems like there are best practices, but it wasn't clear which I should use for this particular app.

Reply:
Per your suggestion, I removed the try catches, which is why this comment became outdated